### ISSUE TITLE
A MACOS version?

### ISSUE THREAD

**Issue by GENEVE0086** [newflasher/newflasher#27]
I am a user of Xperia and MacBook, when I want to flash my phone, I have to use flashtool. But I think newFlasher is better for me because I used to flash my phone on Windows computer, I think a Mac version newFlasher can be good.

**Comment by munjeni** [comment]
Did you tried to compile byself newflasher with macos toolchain?

**Comment by RedL0tus** [comment]
@munjeni I've tried that recently, and it failed with the following information:

```
➜  newflasher git:(master) make -C .
gcc -Wall -O2 newflasher.c -o newflasher -lz -lexpat
newflasher.c:507:9: error: incomplete definition of type 'struct usb_device_descriptor'
        if (dev->idVendor != VID || dev->idProduct != PID)
            ~~~^
newflasher.c:498:9: note: forward declaration of 'struct usb_device_descriptor'
        struct usb_device_descriptor *dev;
               ^
newflasher.c:507:33: error: incomplete definition of type 'struct usb_device_descriptor'
        if (dev->idVendor != VID || dev->idProduct != PID)
                                    ~~~^
newflasher.c:498:9: note: forward declaration of 'struct usb_device_descriptor'
        struct usb_device_descriptor *dev;
               ^
newflasher.c:564:14: warning: implicit declaration of function 'ioctl' is invalid in C99 [-Wimplicit-function-declaration]
                                if ((n = ioctl(fd, USBDEVFS_CLAIMINTERFACE, &ifc)) != 0) {
                                         ^
newflasher.c:564:24: error: use of undeclared identifier 'USBDEVFS_CLAIMINTERFACE'
                                if ((n = ioctl(fd, USBDEVFS_CLAIMINTERFACE, &ifc)) != 0) {
                                                   ^
newflasher.c:826:31: error: variable has incomplete type 'struct usbdevfs_bulktransfer'
        struct usbdevfs_bulktransfer bulk;
                                     ^
newflasher.c:826:9: note: forward declaration of 'struct usbdevfs_bulktransfer'
        struct usbdevfs_bulktransfer bulk;
               ^
newflasher.c:847:9: warning: implicit declaration of function 'ioctl' is invalid in C99 [-Wimplicit-function-declaration]
                                n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                                    ^
newflasher.c:847:24: error: use of undeclared identifier 'USBDEVFS_BULK'
                                n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                                                   ^
newflasher.c:877:8: warning: implicit declaration of function 'ioctl' is invalid in C99 [-Wimplicit-function-declaration]
                        n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                            ^
newflasher.c:877:23: error: use of undeclared identifier 'USBDEVFS_BULK'
                        n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                                           ^
newflasher.c:894:8: warning: implicit declaration of function 'ioctl' is invalid in C99 [-Wimplicit-function-declaration]
                        n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                            ^
newflasher.c:894:23: error: use of undeclared identifier 'USBDEVFS_BULK'
                        n = ioctl(h->desc, USBDEVFS_BULK, &bulk);
                                           ^
newflasher.c:1938:18: warning: implicit declaration of function 'ftello64' is invalid in C99 [-Wimplicit-function-declaration]
        xml_file_size = ftello64(fp);
                        ^
5 warnings and 7 errors generated.
make: *** [newflasher] Error 1
```
macOS seems to be having different API and requires adaptation.

**Comment by munjeni** [comment]
Yes, it need implementation! I have started some work https://github.com/newflasher/newflasher/commit/f43084101af55936279ce6b079faf2ad4632c390 but ended since macos use IOKit/usb/IOUSBLib.h and it need to be backpoorted and I can't do that, hope somebody do that.

**Comment by RedL0tus** [comment]
@munjeni I have experience with neither C nor IOKit, so...

But there are some copy of [macOS SDK available online](https://github.com/phracker/MacOSX-SDKs) (I don't know those are illegal or not), or you may try with CI platforms that supports macOS.

**Comment by munjeni** [comment]
We need to port few things about asynchonous usb transfer functions https://opensource.apple.com/source/IOUSBFamily/IOUSBFamily-281.4.1/IOUSBFamily/Headers/IOUSBLib.h.auto.html , usb_open function and transfer_bulk_async and thats all

**Comment by RedL0tus** [comment]
@munjeni I tried to do that... but I failed because of both my pathetic skill with C and the lack of documentation on IOKit. I'm thinking about... why not use libusb?

**Comment by munjeni** [comment]
Hi, you can use libusb, but situation is the same - needs for backporting initialisation & traansfer functions. Here is my old work based on libusb https://pastebin.com/QKDDRNrc , hope it help you

**Comment by munjeni** [comment]
If you need try MACOS 64 bit build I have done some linusb implementation on this commit https://github.com/newflasher/newflasher/commit/ed10305602e49b779f86cd067c24e967b43ca492 , can't tell you if it will work, it actualy compile, if it actualy work it will need some small additions - modifications.

**Comment by munjeni** [comment]
@RedL0tus any newfls? I have done v22 version which might work on Macos, please test and let me know.

**Comment by RedL0tus** [comment]
@munjeni Hi, I just tried v22.

If I compile newflasher with the bundled libexpat/libz/libusb static library, it will crash.

```
(lldb) target create "newflasher.x86_64-apple-darwin11"
Current executable set to '/Users/kay/Documents/github/newflasher/newflasher.x86_64-apple-darwin11' (x86_64).
(lldb) r
Process 18818 launched: '/Users/kay/Documents/github/newflasher/newflasher.x86_64-apple-darwin11' (x86_64)
--------------------------------------------------------
            newflasher.x86_64-apple-darwin11 v22 by Munjeni @ 2017/2019
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 132219117 MB
  Total space on current drive = 183164297 MB
  Free space on drive          = 136685529 MB
Process 18818 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00007fff319219ff CoreFoundation`CFRetain.cold.1 + 14
CoreFoundation`CFRetain.cold.1:
->  0x7fff319219ff <+14>: ud2

CoreFoundation`__CFGenericTypeID.cold.1:
    0x7fff31921a01 <+0>:  leaq   0x1e9ea4(%rip), %rax      ; "Attempting to address object pointer with invalid tag"
    0x7fff31921a08 <+7>:  movq   %rax, 0x5785bdc1(%rip)    ; gCRAnnotations + 8
    0x7fff31921a0f <+14>: ud2
Target 0: (newflasher.x86_64-apple-darwin11) stopped.
```

Then I tried to compile it with my own libraries, it will not crash but still cannot function correctly:

```
--------------------------------------------------------
            newflasher v22 by Munjeni @ 2017/2019
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 133176689 MB
  Total space on current drive = 183164297 MB
  Free space on drive          = 137643100 MB

Optional step! Type 'y' and press ENTER if you want dump trim area, or type 'n' and press ENTER to skip.
Do in mind this doesn't dump drm key since sake authentifiction is need for that!
y
 - Writing log to tadump.log
 - Writing trim area dump to 01.t
.
 - Error writing command powerdown!
```

```
--------------------------------------------------------
            newflasher v22 by Munjeni @ 2017/2019
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 133180116 MB
  Total space on current drive = 183164297 MB
  Free space on drive          = 137646527 MB

Optional step! Type 'y' and press ENTER if you want dump trim area, or type 'n' and press ENTER to skip.
Do in mind this doesn't dump drm key since sake authentifiction is need for that!
n
 - Error writing command getvar:max-download-size!
 - Error writing command powerdown!
```

I thought it may be some weird permission issue so I tried to run the binary with sudo, but it still shows the same result.

**Comment by munjeni** [comment]
Hi, thanks, try to remove those two lines (479 and 480):

int32_t OSAtomicDecrement32Barrier(volatile int32_t *__theValue) { return 0; }
int32_t OSAtomicIncrement32Barrier(volatile int32_t *__theValue) { return 0; }

I'm hardcoded it since my compiler do not have those two functions.

**Comment by munjeni** [comment]
If not work I will try install mac on my pc in next few days

**Comment by munjeni** [comment]
Seems EP for thasfet out is wrong https://github.com/newflasher/newflasher/blob/master/newflasher.c#L895 , can you connect phone in flash mode and give me all informations you find about usb device, vid, pid, ep, ...etc? Other than that seems vid and pid is succesfully found.

**Comment by munjeni** [comment]
Try put 0x02 to https://github.com/newflasher/newflasher/blob/master/newflasher.c#L895 and let me know

**Comment by RedL0tus** [comment]
@munjeni 

> Hi, thanks, try to remove those two lines (479 and 480):
> 
> int32_t OSAtomicDecrement32Barrier(volatile int32_t *__theValue) { return 0; }
> int32_t OSAtomicIncrement32Barrier(volatile int32_t *__theValue) { return 0; }
> 
> I'm hardcoded it since my compiler do not have those two functions.

The problem still exists after removing these two lines.

> If not work I will try install mac on my pc in next few days.
> 
> Edit:
> Seems I can't install mac without mac? I don't owning mac :(

It is possible, but it requires some effort and some devices do not have drive under macOS. macOS have lots of cursed designs, actually right now I'm thinking it may be not worthy to support macOS, because Apple is constantly breaking compatibility and there are not so many macOS-only users. Maybe just wait for someone who has the interest and the skill to port it if we cannot figure out how to fix it.



> Seems endpoint OUT is wrong https://github.com/newflasher/newflasher/blob/master/newflasher.c#L895 , can you connect phone in flash mode and give me all informations you find about usb device, vid, pid, endpoint, ...etc? Other than that seems vid and pid is succesfully found. I'm unsure now about endpoint IN and OUT, even I'm searching google without luck getting how you can list that info on mac, if you allready know it would be of good help

```
| +-o Android@01300000  <class AppleUSBDevice, id 0x10000ca4d, registered, matched, active, busy 0 (1 ms), retain 14>
  |     {
  |       "sessionID" = 78223957903685
  |       "idProduct" = 45067
  |       "iManufacturer" = 1
  |       "bDeviceClass" = 0
  |       "bMaxPacketSize0" = 64
  |       "bcdDevice" = 1028
  |       "iProduct" = 2
  |       "iSerialNumber" = 0
  |       "bNumConfigurations" = 1
  |       "Bus Power Available" = 250
  |       "USB Address" = 7
  |       "Built-In" = No
  |       "locationID" = 19922944
  |       "bDeviceSubClass" = 0
  |       "bcdUSB" = 512
  |       "USB Product Name" = "Android"
  |       "PortNum" = 3
  |       "non-removable" = "no"
  |       "AppleUSBAlternateServiceRegistryID" = 4295019083
  |       "bDeviceProtocol" = 0
  |       "IOCFPlugInTypes" = {"9dc7b780-9ec0-11d4-a54f-000a27052861"="IOUSBHostFamily.kext/Contents/PlugIns/IOUSBLib.bundle"}
  |       "IOPowerManagement" = {"DevicePowerState"=0,"CurrentPowerState"=3,"CapabilityFlags"=65536,"MaxPowerState"=4,"DriverPowerStat$
  |       "kUSBCurrentConfiguration" = 1
  |       "Device Speed" = 2
  |       "USB Vendor Name" = "Sony"
  |       "idVendor" = 4046
  |       "IOGeneralInterest" = "IOCommand is not serializable"
  |       "IOClassNameOverride" = "IOUSBDevice"
  |     }
```

```
USB:

    USB 3.0 Bus:

      Host Controller Driver: AppleUSBXHCIFL1100
      PCI Device ID: 0x1100
      PCI Revision ID: 0x0010
      PCI Vendor ID: 0x1b73

        Android:

          Product ID: 0xb00b
          Vendor ID: 0x0fce  (Sony Ericsson Mobile Communications AB)
          Version: 4.04
          Speed: Up to 480 Mb/s
          Manufacturer: Sony
          Location ID: 0x01300000 / 7
          Current Available (mA): 500
          Current Required (mA): 500
          Extra Operating Current (mA): 0
```
(I'm using a CalDigit TS3+)



> Try put 0x02 to https://github.com/newflasher/newflasher/blob/master/newflasher.c#L895 and let me know if not work try https://vovkos.github.io/doxyrest/samples/libusb/enum_libusb_request_type.html 0x00 and 0x03

I have just tried `0x00`, `0x02`, `0x03` and the problem persists.

**Comment by munjeni** [comment]
Thanks! I realy have no idea right now, reading this https://vovkos.github.io/doxyrest/samples/libusb/enum_libusb_request_type.html might it be instead e.g. 0x01 something like (0x01 << 5) , you can try if you get more free time.

**Comment by munjeni** [comment]
@RedL0tus I'm made something now hope its work now, let me know please when you get more free time. Compile with your own libusb.

**Comment by RedL0tus** [comment]
@munjeni Sorry for the late reply, I just tried the newest code with my phone:

```
--------------------------------------------------------
            newflasher v24 by Munjeni @ 2017/2019
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 131670245 MB
  Total space on current drive = 183164297 MB
  Free space on drive          = 137447379 MB

Optional step! Type 'y' and press ENTER if you want dump trim area, or type 'n' and press ENTER to skip.
Do in mind this doesn't dump drm key since sake authentifiction is need for that!
n
bulk transfer (out): LIBUSB_ERROR_TIMEOUT
 - Error writing command getvar:max-download-size!
bulk transfer (out): LIBUSB_ERROR_TIMEOUT
 - Error writing command powerdown!
```

I also tried with sudo but got the same result.

**Comment by munjeni** [comment]
No prob, I like in any time! First of all many thanks! I'm done much more work, https://github.com/newflasher/newflasher/commit/e1df0bcdcd5bf20b1d29145dddd8721d9ce33be6 , can this work now?

**Comment by RedL0tus** [comment]
@munjeni 

```
--------------------------------------------------------
            newflasher v24 by Munjeni @ 2017/2019
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 131857542 MB
  Total space on current drive = 183164297 MB
  Free space on drive          = 137110482 MB

Device properties:
        bus number: 1
         port path: 3 (from root hub)
             speed: 480 Mbit/s (USB HighSpeed)

Reading device descriptor:
            length: 18
      device class: 0
               S/N: 0
           VID:PID: 0xFCE:0xB00B
         bcdDevice: 0x404
   iMan:iProd:iSer: 1:2:0
          nb confs: 1

Reading configuration descriptors:
             nb interfaces: 1
              interface[0]: id = 0
interface[0].altsetting[0]: num endpoints = 2
   Class.SubClass.Protocol: 0xFF 0x42 0x3
       endpoint[0].address: 0x81
           max packet size: 0x200
          polling interval: 0x0
       endpoint[1].address: 0x01
           max packet size: 0x200
          polling interval: 0x0

Claiming interface 0...

Reading string descriptors:
   String (0x01): "Sony"
   String (0x02): "Android"

Optional step! Type 'y' and press ENTER if you want dump trim area, or type 'n' and press ENTER to skip.
Do in mind this doesn't dump drm key since sake authentifiction is need for that!
n
bulk transfer (out): LIBUSB_ERROR_NOT_FOUND
 - Error writing command getvar:max-download-size!
bulk transfer (out): LIBUSB_ERROR_NOT_FOUND
 - Error writing command powerdown!
```

Hmmmmmmm...

**Comment by munjeni** [comment]
Thanks! I don't have idea right now but I'm in contact with libusb guys https://github.com/newflasher/newflasher/commit/fde5c13d7a90beaecaaa1a13081204833c5fb647 hope we address isue. On google search seems a lot of thar LIBUSB_ERROR_NOT_FOUND results

**Comment by munjeni** [comment]
I will try to install mac https://isoriver.com/how-to-install-mac-os-on-windows-pc-using-virtual-machine/

**Comment by munjeni** [comment]
@RedL0tus I'm succed installing mac, I'm fixed some isues related to ep_in and ep_out, but still unable to get bulk transfer working on vmware, getting some kernel error, my hope it is related to vmware, can you try now when you get free time?

**Comment by munjeni** [comment]
@RedL0tus got it!!! \0/ If you going to test please do not flash .ta. files, partition , boot delivery, bootloader ...etc, you can try for the start with cache partition, if flashing cache suced try to flash system, do not try flash anything else since I unsure now if bulk transfer will work on larger chunks, if booth cache and system flash sucesfully the rest might work. Do in mind this is alpha support for macos, it might work, it might crash, need to be tested

**Comment by RedL0tus** [comment]
@munjeni 

That's great! I just tried with cache.sin and it seems to be successfully flashed, but I got another LIBUSB_ERROR_TIMEOUT when newflasher trying to put my phone out of flash mode.

```
Device is put now out of flash mode.
Sent command: Sync
bulk transfer (in): LIBUSB_ERROR_TIMEOUT
 - Error reply: null!
Error, no sync response!
```

Right now this phone is no longer used as my main device and I'm willing to help more on testing.

**Comment by munjeni** [comment]
I'm tested on trim area dump and it suced too, bufer exceeds 0x200 bytes, so seems everything is ok. About sync response bug, it was reported some years ago, see https://github.com/newflasher/newflasher/commit/616aca11cbdd8e8947d60ff3d2e2372bfdb67e17 I'm temporary enabled it but seems not all phones giving reply :( Right now that error can be ignored since sync is sucesfully writed and thats important, the rest after sync is a less important but I will reopen isue hope we solve that soon. You can fully flash test now? Try first with system , two - three times if need, than try slowly the rest, the last flash bootdelivery and ta files, I'm believing everything is right.

**Comment by munjeni** [comment]
What is your mac version? I'm testing right now on 10.14

**Comment by RedL0tus** [comment]
@munjeni 

> I'm tested on trim area dump and it suced too, bufer exceeds 0x200 bytes, so seems everything is ok. About sync response bug, it was reported some years ago, see [616aca1](https://github.com/newflasher/newflasher/commit/616aca11cbdd8e8947d60ff3d2e2372bfdb67e17) I'm temporary enabled it but seems not all phones giving reply :( Right now that error can be ignored since sync is sucesfully writed and thats important, the rest after sync is a less important but I will reopen isue hope we solve that soon. You can fully flash test now? Try first with system , two - three times if need, than try slowly the rest, the last flash bootdelivery and ta files, I'm believing everything is right.

I'm downloading the system.sin file from my NAS, I'll start test them soon.



> What is your mac version? I'm testing right now on 10.14

I'm running macOS 10.15.5 on a MacBook Pro 13 2019.

**Comment by munjeni** [comment]
Ok, many thanks! It might be that sync not fail all the time? I will test it more soon.

**Comment by munjeni** [comment]
Just tested, for me sync is ok

Device is put now in flash mode.

No .sin files in partition dir...
You must extract partition.zip into 'partition' folder if you want flash partition image!
On 2018 and UP models you must move partition sin files to 'partition' folder if you need flash partition images!

Processing .sin files...
Created ouput folder flash_session

Processing appslog_X-FLASH-ALL-8F02.sin
 - Extracting from appslog_X-FLASH-ALL-8F02.sin
 - Extracting signature appslog.cms
 - Uploading signature ./flash_session/appslog.cms
      signature:00000533
      OKAY.
 - Extracting sparse chunk appslog.000
 - Uploading sparse chunk ./flash_session/appslog.000
      download:000003c2
      OKAY.
      erase:appslog
      OKAY.
      flash:appslog
      OKAY.
 - End of appslog_X-FLASH-ALL-8F02.sin

Processing .ta files...
No .ta files in current dir.

Processing boot delivery...

boot_delivery.xml not exist in boot folder or no boot folder.

Set slot 'b' active.

Device is put now out of flash mode.
Sent command: Sync
Sent command: reboot.

Done.
Releasing interface 0.
Closing device.
macoss-Mac:newflasher macos$

**Comment by munjeni** [comment]
@RedL0tus I'm believing Sync isue is caused by not waiting at least some secconds for sync to finish, I'm modified sync command now, can you try now?

**Comment by munjeni** [comment]
I have flashed whole firmware, all seems right, no sync isue, no isues at all, need somebody confirm.

**Comment by RedL0tus** [comment]
@munjeni 

> I have flashed whole firmware with v29 version (latest), all seems right, no sync isue, no isues at all, need somebody confirm.

I tried to flash the system.sin and it works flawlessly. I did not met the sync issue this time, I think it could be a firmware-related bug as my phone is a XZ1C.

**Comment by munjeni** [comment]
Many thanks for your testing time, since thing is confirmed I'm closing two tickets, two things seems fixed, if you notice any isue please reopen, I think thats it, there will be no more sync isues.

**Comment by RedL0tus** [comment]
Thank you for your work!

**Comment by ZegWe** [comment]
hi, I tried to use newflasher on my Mac to flash my xperia 1 iii, but something seems went wrong, nothing flashed into my phone, and the phone stayed in flash mode until I reboot manually.
p.s. libusb had upgrade to version 1.0.24
```
--------------------------------------------------------
            newflasher v51 by Munjeni @ 2017/2020           
--------------------------------------------------------

Determining available free space:

  Available space to caller    = 34546642 MB
  Total space on current drive = 122086923 MB
  Free space on drive          = 34546642 MB

Device properties:
        bus number: 20
         port path: 4 (from root hub)
             speed: 5000 Mbit/s (USB SuperSpeed)

Reading device descriptor:
            length: 18
      device class: 0
               S/N: 0
           VID:PID: 0xFCE:0xB00B
         bcdDevice: 0x100
   iMan:iProd:iSer: 1:2:0
          nb confs: 1

Reading configuration descriptors:
             nb interfaces: 1
              interface[0]: id = 0
interface[0].altsetting[0]: num endpoints = 2
   Class.SubClass.Protocol: 0xFF 0x42 0x3
       endpoint[0].address: 0x81
           max packet size: 0x400
          polling interval: 0x0
       endpoint[1].address: 0x01
           max packet size: 0x400
          polling interval: 0x0

Set configuration: succed.

Claiming interface 0: succed.

Reading string descriptors:
   String (0x01): "Google"
   String (0x02): "Android"


Reboot mode at the end of flashing:
  typa 'a' for reboot to android, type 'f' for reboot to fastboot, type 's' for reboot to same mode, type 'p' for poweroff, and press ENTER.
a

Optional step! Type 'y' and press ENTER if you want dump trim area, or type 'n' and press ENTER to skip.
Do in mind this doesn't dump drm key since sake authentifiction is need for that! But it is recommend to have dump in case hard brick!
n
bulk transfer (out): LIBUSB_ERROR_TIMEOUT
 - Error writing command getvar:slot-count, ignore this error!
bulk transfer (out): LIBUSB_ERROR_TIMEOUT
 - Error writing command getvar:Battery!
bulk transfer (out): LIBUSB_ERROR_TIMEOUT
 - Error writing command continue!
Releasing interface 0.
Closing device.
Saving session...completed.
```

**Comment by munjeni** [comment]
Hi, on Darwin you need to install libusb-1.0 otherwise it not going to work, don't ask why I don't know
